Iterate on Java codegen

jmchilton opened the pull request:
Build maven project containing Java objects for record types, capable of validation, and that can generate JavaDocs with schema docs contained within. 

Missing stuff tracked #303.

tetron commented:
One thing you will want to figure out early on is how best to represent sum types in Java, eg a `run` field can be an embedded Workflow or a CommandLineTool or a ExpressionTool or a URI string.  Returning `Object` and casting everywhere is going to get tedious very quickly.

jmchilton commented:
Seems like Object would at least be a place to start 😅. The yak I'm shaving with this project currently I think would mostly just benefit from the UnionLoader loading types and verifying things. Validation still happens even if the interface to using the beans is crap.

Still... if I were going to put time into this... hmm... my first thought is it may be worth building UnionOf2<T1, T2>, UnionOf3<T1, T2, T3> .... and UnionOf2Loader, UnionOf3Loader, ....

```
class UnionOf2<T1, T2> {
   private Object value;
   public Object get();
   private int index; // track type of object by index 
   public boolean is1();
   public boolean is2();
   public T1 get1();
   public T2 get2();
   public Optional<T1> getOptional1();
   public Optional<T2> getOptional2();
}
```

The type definitions on the beans could become very ugly but may be somewhat informative. Certainly the potential for a lot more type safety also.

Nervous about investing too much time here though.

tetron commented:
@jmchilton you're a beast, thank you so much for working on this

jmchilton commented:
I got tripped up because schema-salad-tool and codegen modules treat fields with ``default`` but not explicitly annotated as optional differently. Probably CWL specs were better about this and so it hasn't been encountered. Eliminating those from my schema - this works! It really validates my relatively complex schema.

jmchilton commented:
Now with encoded Java formatting (https://github.com/google/google-java-format) and JavaDocs.

<img width="1669" alt="Screen Shot 2019-11-27 at 9 47 16 PM" src="https://user-images.githubusercontent.com/216771/69773039-b4235080-115f-11ea-8b72-54c84f5125c9.png">

jmchilton commented:
<img width="1018" alt="Screen Shot 2019-11-28 at 10 02 36 AM" src="https://user-images.githubusercontent.com/216771/69816386-452d1280-11c6-11ea-9d63-c90c4287036d.png">

mr-c commented:
@jmchilton Super cool! Can you add a `main` method?

mr-c commented:
And while I'm making wishes, it would be nice if the schema docs became javadocs too.

I tried your code with both cwl v1.0 and v1.1 's schemas, no crashes :-)

jmchilton commented:
A main method that validates a document I assume? Sounds easy enough.

Placing the schema docs in the javadocs is a cool idea! I'm not sure...

jmchilton commented:
```shell
$ mvn assembly:single
$ java -jar target/gxformat2-0.0.1-SNAPSHOT-jar-with-dependencies.jar ../test-workflow.yml
Exception in thread "main" org.galaxyproject.gxformat2.gxformat2.utils.ValidationException: Failed to match union type
  Trying 'RecordField'
    the `inputs` field is not valid because:
      Failed to match union type
        Expected object with Java type of java.util.List but got java.util.HashMap
        Trying 'RecordField'
          the `type` field is not valid because:
            Failed to match union type
              Expected null
              Expected one of [File, data, collection]
  Expected object with Java type of java.util.List but got java.util.LinkedHashMap
	at org.galaxyproject.gxformat2.gxformat2.utils.UnionLoader.load(UnionLoader.java:31)
	at org.galaxyproject.gxformat2.gxformat2.utils.Loader.documentLoad(Loader.java:41)
	at org.galaxyproject.gxformat2.gxformat2.utils.RootLoader.loadDocument(RootLoader.java:19)
	at org.galaxyproject.gxformat2.gxformat2.utils.RootLoader.loadDocument(RootLoader.java:85)
	at org.galaxyproject.gxformat2.gxformat2.utils.RootLoader.loadDocument(RootLoader.java:73)
	at org.galaxyproject.gxformat2.gxformat2.utils.RootLoader.loadDocument(RootLoader.java:65)
	at org.galaxyproject.gxformat2.gxformat2.utils.RootLoader.loadDocument(RootLoader.java:31)
	at org.galaxyproject.gxformat2.gxformat2.utils.RootLoader.loadDocument(RootLoader.java:48)
	at org.galaxyproject.gxformat2.gxformat2.utils.Validator.main(Validator.java:12)
```

mr-c commented:
> I registered the requested validator as the entry point for the jar and I added the maven assembly plugin to build a self contained jar variant that is good for a standalone demo

Score!

mr-c commented:
hmm.. maybe I have the wrong version of java and/or maven?
```
$ schema-salad-tool --codegen java --codegen-target metaschema/ schema_salad/metaschema/metaschema.yml
metaschema $ java -jar target/salad-0.0.1-SNAPSHOT-jar-with-dependencies.jar ~/cwl-v1.1/CommonWorkflowLanguage.yml 
Error: Could not find or load main class org.w3id.cwl.salad.utils.Validator
Caused by: java.lang.ClassNotFoundException: org.w3id.cwl.salad.utils.Validator
$ mvn --version
Apache Maven 3.6.0
Maven home: /usr/share/maven
Java version: 11.0.4, vendor: Ubuntu, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.3.0-23-generic", arch: "amd64", family: "unix"
```

jmchilton commented:
Good catch - looks like you need to call something to compile the stuff before you call the thing to assemble the stuff. I assumed ``mvn assembly:single`` would know what needs to happen before that. ``mvn compile`` followed by that ensured the correct class file was in the jar for me - ``mvn install`` might be even safer? I'll see what I can come up with and update the generated README.

I'd like to dispatch with these type problems - the big one is my subclassing TypeDef to JavaTypeDef. If this were Java - I'd parameterize the codegen code ``<? extends TypeDef>`` but I don't know how to do that in Python? I throw the new params in a ``Dict[Text, Text]`` but that would be losing type info to get the type system to eat the code. I could also just place the new parameters on ``TypeDef`` I suppose? It isn't the cleanest but I guess there is no problem per se? Any advice?

jmchilton commented:
Also any clue what to do about the type system thinking ``resource_listdir`` is present. It is in the typeshed file commented out:

```
# Names in __all__ with no definition:
#   add_activation_listener
#   cleanup_resources
#   iter_entry_points
#   resource_exists
#   resource_filename
#   resource_isdir
#   resource_listdir
#   resource_stream
#   resource_string
#   set_extraction_path
```

jmchilton commented:
Progress today - I implemented a test case - keep regressing on one Python or the other.

Explored typing because I'm a nerd. 

Optional types (``foo?`` becomes ``java.util.Optional<Foo>`` instead of just a ``Object``) are I think a success and I'm proud of them, as are converting schema salad enums to Java enums:

<img width="1643" alt="Screen Shot 2019-11-29 at 5 05 14 PM" src="https://user-images.githubusercontent.com/216771/69891432-742cac80-12ca-11ea-9578-bdf5bc3e2716.png">
<img width="1655" alt="Screen Shot 2019-11-29 at 5 05 06 PM" src="https://user-images.githubusercontent.com/216771/69891434-77279d00-12ca-11ea-961e-72dc00e99dfd.png">

I wanted the interface for the Galaxy workflow language to be better - the problem is the root loader at the top returns either a GalaxyWorkflow or a List of Objects. So there was no good interface. I tried implementing this class ``OneOrListOf`` with corresponding typed loader. It provides better validation, docs, and typing for my language:

<img width="1644" alt="Screen Shot 2019-11-29 at 5 07 58 PM" src="https://user-images.githubusercontent.com/216771/69891475-d2f22600-12ca-11ea-851b-f45d9320897e.png">

The problem is that it doesn't do this for CWL or the meta schema:

<img width="1653" alt="Screen Shot 2019-11-29 at 5 08 52 PM" src="https://user-images.githubusercontent.com/216771/69891489-f2894e80-12ca-11ea-9053-61ecd72343aa.png">

Presumably this is because the thing at the top is polymorphic so the corresponding loader is more complicated.

So while I think the interface is an improvement for simple schema - I don't know that it will be worth it in general. Would be odd to have some random part of the schema that follows that pattern warp into this custom object but not other parts. 

So we might want to revert that comment or spend more time thinking about polymorphism and capture the parent superclass of such loaders.

jmchilton commented:
Wasn't my favorite design decision but I resolved a bunch of the type problems by just merging the JavaTypeDef changes back into TypeDef as optional attributes. Fixed most of the remaining type issues.

Added parts of the schema salad docs into the JavaDocs - class, interface, and getters:

<img width="1645" alt="Screen Shot 2019-12-01 at 11 40 21 PM" src="https://user-images.githubusercontent.com/216771/69931364-008acb00-1495-11ea-8d4a-b83d277875dd.png">

mr-c commented on schema_salad/main.py:
```diff
@@ -105,6 +105,21 @@ def main(argsl=None):  # type: (Optional[List[str]]) -> int
+        "--codegen-target",
+        type=str,
+        default=None,
+        help="Defaults to sys.stdout for python (and ./ for hidden Java codegen)",
```
Hidden?

jmchilton replied:
"Generate classes in target language, currently supported: python"

I'll add java to the list and remove the word hidden in a follow up.

mr-c commented on schema_salad/main.py:
```diff
@@ -105,6 +105,21 @@ def main(argsl=None):  # type: (Optional[List[str]]) -> int
+        type=str,
+        metavar="directory",
+        default=None,
+        help="Directory of example documents for test case generation (Java only).",
```
Is there an example of how to use this?

jmchilton replied:
https://github.com/common-workflow-language/schema_salad/blob/master/schema_salad/tests/test_java_codegen.py#L23

I wanted to allow valid and invalid document tests but I never got that far. Currently just assumes all the files in the examples directory are valid documents of this type.

mr-c commented:
W00t! Shall I merge?

jmchilton commented:
I have followups in the pipeline - but I'm not touching anything while this is such a pretty shade of green 😆. I'd go ahead an merge now.
